src/query_plan/fetch_dependency_graph.rs

```diff
@@ -144,3 +144,67 @@ pub(crate) struct FetchDependencyGraphPath {
     path_in_node: OpPath,
     response_path: Vec<FetchDataPathElement>,
 }
+
+impl FetchDependencyGraph {
+    pub(crate) fn new(
+        supergraph_schema: ValidFederationSchema,
+        federated_query_graph: Arc<QueryGraph>,
+        root_type_for_defer: Option<CompositeTypeDefinitionPosition>,
+        starting_id_generation: u64,
+    ) -> Self {
+        Self {
+            defer_tracking: DeferTracking::empty(&supergraph_schema, root_type_for_defer),
+            supergraph_schema,
+            federated_query_graph,
+            graph: Default::default(),
+            root_nodes_by_subgraph: Default::default(),
+            starting_id_generation,
+            fetch_id_generation: starting_id_generation,
+            is_reduced: false,
+            is_optimized: false,
+        }
+    }
+
+    /// Must be called every time the "shape" of the graph is modified
+    /// to know that the graph may not be minimal/optimized anymore.
+    fn on_modification(&mut self) {
+        self.is_reduced = false;
+        self.is_optimized = false;
+    }
+
+    pub(crate) fn get_or_create_root_fetch_group(
+        &mut self,
+        subgraph_name: &NodeStr,
+        root_kind: SchemaRootDefinitionKind,
+        parent_type: CompositeTypeDefinitionPosition,
+    ) -> &IndexSet<NodeIndex> {
```

Comment:
This function should be returning the equivalent of a `FetchGroup`, which is a node in the fetch dependency graph (so I think a `NodeIndex` for a node that is either fetched or created by this function).